[WIP][11.0][ADD] mobile app picking

legalsylvain opened the pull request:
This PR adds a new module ``mobile_app_picking`` with similar technology as ``mobile_app_inventory``. (light web application, based on angular framework, in the same repository).

a refactor has been done, to split the module into three modules : 
    * ``mobile_app_abstract`` = base module with generic API tools 
    * ``mobile_app_angular`` = base module with angular framework and the custom lib odoo_angular.js
    * ``mobile_app_picking`` = business module that allow user to check product quantities and confirm pickings). 

This refactor was mentionned [here](https://github.com/OCA/stock-logistics-barcode/tree/8.0/mobile_app_inventory#roadmap--current-limits)

See the full readme description with screenshots [here](https://github.com/legalsylvain/stock-logistics-barcode/blob/11.0_ADD_mobile_app_picking/mobile_app_picking/README.rst)


Note : the readme file will be incorrectly generated, as long as https://github.com/OCA/maintainer-tools/pull/404 is not merged.

legalsylvain committed:
[ADD] mobile_app_picking
[REF] based on mobile_app_inventory, split into 3 modules. mobile_app_abstract (API), mobile_app_angular (JS Framework), mobile_app_picking
[REF] remove useless code for odoo 7 in odoo_angular.js
[REF] remove sync functions in odoo_angular.js
[ADD] eslintrc.js file

robinkeunen commented on mobile_app_picking/static/www/controllers/main_scan.js:
```diff
@@ -0,0 +1,86 @@
+            $stateParams.picking_id,
+            inputValue).then(function (moveLine) {
+            if (!moveLine) {
+              $scope.display_loading_end('Unable to found the barcode ' + inputValue)
```
typo: unable to **find\***

legalsylvain committed:
[V1.1]

robinkeunen reviewed (commented):
erreur anglais. Je te fais une vraie review dans la semaine.

JayVora-SerpentCS commented:
Looks so good

legalsylvain committed:
version 1.2
[ADD] controller to have a shortcut for the route /mobile_app_picking
[ADD] checkbox to display only 'todo' move lines.
[FIX] display product variant name instead of product template name.
[FIX] css + UI button texts.
[ADD] new text if picking types / picking list are empty.

Joaco1980 commented:
The app does not work with the multi_ean module. It only works with the main ean, if it has more barcodes it does not read well.

legalsylvain committed:
[V1.3]
[ADD] new page to handle confirmation, during picking validation.

legalsylvain commented:
> The app does not work with the multi_ean module. It only works with the main ean, if it has more barcodes it does not read well.

Yes the module ``multi_ean`` is an extra addons (out of the Odoo Core), and most of the modules are not working correctly when it is installed, including, AFAIK, core modules like ``point_of_sale``.

Feel free to develop a glue module to make both modules working together.

Commit:
[IMP] add link to open picking in mobile app

legalsylvain committed:
[REF] use moves instead of move lines
[FIX] eslint
[DOC] add confirmation description
[DOC] improve roadmap usage

legalsylvain committed:
[DOC] set modules in Beta development status

legalsylvain committed:
[ADD] test

agent-z28 commented:
Respect.

rousseldenis commented:
LGTM. Basic functional review. It does the job.

Apart of this, I believe that scanning must obey flexible scenarios (but user-constrained) and I never encountered better solution than stock_scanner.

There are pending developments to make a robust stock_scanner_web.

@legalsylvain Did you see https://docs.google.com/document/d/1mct6bFFWJqW01wGFcjc-uQNEjyCxvh6Y9TuFdRhe-b0 from @jgrandguillaume as it talks about stock_scanner  ?

pedrobaeza commented:
We are also developing in September the picking part of `stock_barcodes`.

robinkeunen commented:
A client of ours has been using `mobile_app_picking` for 5 months in production now. He's really happy about it. Fast and user friendly.

agent-z28 commented:
> We are also developing in September the picking part of `stock_barcodes`.

@pedrobaeza 

We also did one draft, now we found this and wanted to merge our solutions, or complete this solution with creating incoming and outgoing.. 

What exactly are you planing to develop? I know the mentioned document.
Doesnt it make sense to resolve those requirements in this angular base?

agenterp commented:
> LGTM. Basic functional review. It does the job.
> 
> Apart of this, I believe that scanning must obey flexible scenarios (but user-constrained) and I never encountered better solution than stock_scanner.
> 
> There are pending developments to make a robust stock_scanner_web.
> 
> @legalsylvain Did you see https://docs.google.com/document/d/1mct6bFFWJqW01wGFcjc-uQNEjyCxvh6Y9TuFdRhe-b0 from @jgrandguillaume as it talks about stock_scanner ?

What do you think of using this as base and adding the stock_scanner features in this angular framework

pedrobaeza commented:
We don't want extra technologies to maintain. The mentioned module does everything inside Odoo web client, but with mobile + barcode scanner facilities.

agent-z28 commented:
> We don't want extra technologies to maintain. The mentioned module does everything inside Odoo web client, but with mobile + barcode scanner facilities.

@pedrobaeza
Is there already somehthing that we could see and help?
We need to define our Picking App Strategy, if we go with this angular, our own, or the "OCA" one.
Would be great to know, also we could help!

agent-z28 commented:
@legalsylvain 
I have one question:
As it looks like, in the form view, the user cannot scan directly.
First he has to press "Scanner". Is this intentionally? 

What do you think of directly fetching the scanned code and process it?

E.G.Move the Blue Input Field into the Form View, so the whole "Scanner View" will be obsolete.

![Bildschirmfoto vom 2019-09-20 10-57-07](https://user-images.githubusercontent.com/7473428/65314028-c1d3bd00-db95-11e9-82dd-96524a8a7f8a.png)

![Bildschirmfoto vom 2019-09-20 10-57-19](https://user-images.githubusercontent.com/7473428/65314038-c6987100-db95-11e9-9ca5-b575703ca528.png)

pedrobaeza commented:
@agenterp you already have it in https://github.com/OCA/stock-logistics-barcode/tree/11.0/stock_barcodes

legalsylvain committed:
[DOC] improve roadmap and description after reviews

legalsylvain commented:
Hi. Thanks for your review.

> [...] E.G.Move the Blue Input Field into the Form View, so the whole "Scanner View" will be obsolete.

Indeed, but not very easy to do it. And adding too much thing in a single screen make it not usable on mobile. (that is the objective of this module). I added a text in the roadmap section.

Also, I added a section in the description, making the promotion of the other approachs. (``stock_scanner`` and ``stock_barcodes``)

/ocabot merge

agent-z28 commented:
legalsylvain,
after long, really long trying and playing with allllll possible ways. 
We migrated the old v8 picking app, then we created an own in odoo app, then we checked OCA APP Pedro suggested. 
Now we finally decided to use your approach, It's fu**** fast and customizeable.
We already did some changes and will add some pull requets soon!
Great WORK!

legalsylvain committed:
[FIX] pylint

legalsylvain commented:
Hi @agenterp : Thanks a lot ! 
Well, I make the changes and merge to have the possibility to port the module into V12.
In the first step, the two framework modules ``mobile_app_abstract`` and ``mobile_app_angular`` + 
- port and make dependency for ``mobile_app_inventory`` (OCA https://github.com/OCA/stock-logistics-barcode/tree/8.0/mobile_app_inventory)
- share to the OCA a GRAP module in incubator repository ``mobile_app_purchase`` : https://github.com/grap/grap-odoo-incubator/tree/8.0/mobile_app_purchase)

And I'll not work for the time being, in ``mobile_app_picking``.

What changes did you make ? only on ``mobile_app_picking`` module ? or on the "framework" modules ?
if it is on the framework modules, could you share you changes quickly, to avoid I port a non optimized module ? 
thanks !

legalsylvain commented:
@OCA-git-bot : could fix trivial pylint error ? ;-)

/ocabot merge

agent-z28 commented:
@legalsylvain we only will change mobile_app_picking. we will submit a pull request within 2 weeks
